beater/java_attacher/java_attacher.go

```diff
@@ -0,0 +1,120 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package javaattacher
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/elastic/apm-server/beater/config"
+)
+
+type JavaAttacher struct {
+	cfg    config.JavaAttacherConfig
+	logger *logp.Logger
+}
+
+func New(cfg config.JavaAttacherConfig) (JavaAttacher, error) {
+	if cfg.JavaBin == "" {
+		if jh := os.Getenv("JAVA_HOME"); jh != "" {
+			cfg.JavaBin = filepath.Join(jh, "/bin/java")
+		} else {
+			bin, err := exec.Command("which", "java").Output()
+			if err != nil {
+				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+			}
+			cfg.JavaBin = string(bin)
+		}
+	}
+	logger := logp.NewLogger("java-attacher")
+	return JavaAttacher{cfg: cfg, logger: logger}, nil
+}
+
+// javaAttacher is bundled by the server
+// TODO: Figure out the real path
+const javaAttacher = "/bin/apm-agent-attach-cli-1.24.0-slim.jar"
+
+func (j JavaAttacher) Run(ctx context.Context) error {
+	cmd := j.build(ctx)
+	j.logger.Debugf("starting java attacher with command: %s", strings.Join(cmd.Args, " "))
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// TODO: How do we want to handle logging?
+	// See comment:
+	// https://github.com/elastic/apm-server/issues/4830#issuecomment-863207642
```

Comment:
something to keep in mind is that we need to ensure that attacher logs are ECS compatible, because we can't mix ECS and non ECS when ECS is enabled (oh well, how many times can i say "ECS" in sentence?)

Reply:
hm, maybe @felixbarny knows?

Reply:
The attacher logs are ECS compatible.